skglm/solvers/group_prox_newton.py

```diff
@@ -0,0 +1,357 @@
+            ws = np.argpartition(opt, -ws_size)[-ws_size:]  # k-largest items (no sort)
+
+            grad_ws = _slice_array(grad, ws, grp_ptr, grp_indices)
+            tol_in = EPS_TOL * stop_crit
+
+            # solve subproblem restricted to ws
+            for pn_iter in range(self.max_pn_iter):
+                # find descent direction
+                delta_w_ws, X_delta_w_ws = _descent_direction(
+                    X, y, w, Xw, fit_intercept, grad_ws, datafit, penalty,
+                    ws, tol=EPS_TOL*tol_in)
+
+                # find a suitable step size and in-place update w, Xw
+                grad_ws[:] = _backtrack_line_search(
+                    X, y, w, Xw, fit_intercept, datafit, penalty,
+                    delta_w_ws, X_delta_w_ws, ws)
+
+                # check convergence
+                opt_in = penalty.subdiff_distance(w, grad_ws, ws)
+                stop_crit_in = np.max(opt_in)
+
+                # optimality of intercept
+                if fit_intercept:
+                    # gradient w.r.t. intercept (constant features of ones)
+                    intercept_opt_in = np.abs(np.sum(datafit.raw_grad(y, Xw)))
+                else:
+                    intercept_opt_in = 0.
+
+                stop_crit_in = max(stop_crit_in, intercept_opt_in)
+
+                if max(self.verbose-1, 0):
+                    p_obj = datafit.value(y, w, Xw) + penalty.value(w[:n_features])
+                    print(
+                        f"PN iteration {pn_iter+1}: {p_obj:.10f}, "
+                        f"stopping crit in: {stop_crit_in:.2e}"
+                    )
+
+                if stop_crit_in <= tol_in:
+                    if max(self.verbose-1, 0):
+                        print("Early exit")
+                    break
+
+            p_obj = datafit.value(y, w, Xw) + penalty.value(w[:n_features])
+            p_objs_out.append(p_obj)
+        return w, np.asarray(p_objs_out), stop_crit
+
+
+@njit
+def _descent_direction(X, y, w_epoch, Xw_epoch, fit_intercept, grad_ws, datafit,
+                       penalty, ws, tol):
+    # given:
+    #   1) b = \nabla F(X w_epoch)
+    #   2) D = \nabla^2 F(X w_epoch)  <------>  raw_hess
+    # minimize quadratic approximation for delta_w = w - w_epoch:
+    #  b.T @ X @ delta_w + \
+    #  1/2 * delta_w.T @ (X.T @ D @ X) @ delta_w + penalty(w)
+    # In BCD, we leverage inequality:
+    #  penalty_g(w_g) + 1/2 ||delta_w_g||_H <= \
+    #  penalty_g(w_g) + 1/2 * || H || * ||delta_w_g||
+    grp_ptr, grp_indices = penalty.grp_ptr, penalty.grp_indices
+    n_features_ws = sum([penalty.grp_ptr[g+1] - penalty.grp_ptr[g] for g in ws])
+    raw_hess = datafit.raw_hessian(y, Xw_epoch)
+
+    lipchitz = np.zeros(len(ws))
+    for idx, g in enumerate(ws):
+        grp_g_indices = grp_indices[grp_ptr[g]:grp_ptr[g+1]]
+        # norm(X[:, grp_g_indices].T @ np.diag(raw_hess) @ X[:, grp_g_indices], ord=2)
+        lipchitz[idx] = norm(_diag_times_X_g(
```

Comment:
I guess you encapsulated the `_diag_times_X_g` into a function to isolate the jit, this might worth a comment

Reply:
I made a ``_diag_times_X_g`` function to avoid copying the columns of ``X``.
The same operation can be achieved
```python
norm(X[:, grp_g_indices] @ np.sqrt(raw_hess), ord=2)**2
```
I made a comment about that in the function definition
https://github.com/scikit-learn-contrib/skglm/blob/99482daf44739bf9337992a278067386185049a5/skglm/solvers/group_prox_newton.py#L350-L353

Reply:
thx!